### PR TITLE
Memory optimizations

### PR DESCRIPTION
**Pull request by achouippe**
This PR tries to reduce the memory cost of Neurow subscriptions.

**TL;DR** 
By hibernating SSE monitors and manually running the garbage collector in SSE processes, the PR decreases the memory cost of SSE subscriptions from about 63ko per subscription to 30ko per subscription. In a server that handles 80k SSE subscription, it should save about **2.5Go of RAM**. This optimization should however slightly increase the CPU usage because garbage collection run more often.

Note: This PR not actually change the max memory required by a Neurow server. The server would eventually run garbage collection in order to free memory to re-use it just after. It just release the memory without having to wait for a gc when the server needs more RAM. This makes its behavior more predictable, and allows to set auto-scaling rules based on the actual RAM usage.

## Analysis
The module `Neurow.Stats.Processes` analyses the process memory usage: It groups processes by name and count their memory cost, number of process and number of messages.

It can be used in the IEX console by calling `Neurow.Stats.Processes.grouping_stats`, or by calling `GET /process_stats` on the internal API.

Locally, by starting 2k concurrent subscriptions that subscribe to 10 shared topics, and then by continuously sending messages on these 10 topics, here are some interesting outliers:

| name / initial function | current function | process count | memory | cost per subscription (ko) | cost for 80k subscriptions (mo)
| -- | -- | -- | -- | -- | -- |
cowboy_stream_h:request_process/3 | Elixir.Neurow.PublicApi.Endpoint:loop/7 | 2000 | 80791512 | 39.45 | 3,081.95
cowboy_clear:connection_process/4 | cowboy_http:loop/1 | 2000 | 43334224 | 21.16 | 1,653.07
Elixir.Neurow.PublicApi.SSEMonitor:init/1 | gen_server:loop/7 | 2000 | 5561152 | 2.72 | 212.14
  |   |   |   |   |  
  |   |   | **Total** | **63.32** | **4,947.16** |

No issue were found in the number of messages.

### Hibernate SSE Monitors
Erlang allows to manually put processes in [hibernation](https://www.erlang.org/doc/apps/erts/erlang.html#hibernate/3):
> Puts the calling process into a wait state where its memory allocation has been reduced as much as possible. This is useful if the process does not expect to receive any messages soon.

SSEMonitor is a genserver, putting it in hibernation is simple: It just requires to return `{:ok, state, :hibernate}` from the init function.

### Manual garbage collection in SSE subscriptions
Because SSE requests handlers are not genservers, it is much more complex to hibernate them. 

`:erlang.hibernate` exists and use it with `Process.send_after` could allow to implement a ticker, but `:erlang.hibernate` erases the process call stack. That means all instructions that should run after that `Neurow.PublicApi.Endpoint#subscribe` returns will actually not run if the SSE process hibernates.

So instead of hibernating we just first force a garbage collection during each SSE passive loop. Compare to an actual  hibernation the call stack is not deleted, and all live data are not moved to a continuous heap. But it still provides significant improvements.

### Results

| name / initial function | Current function | process count | memory | cost per subscription (Ko) | cost for 80k subscriptions (Mo)
| -- | -- | -- | -- | -- | -- |
cowboy_stream_h:request_process/3 | Elixir.Neurow.PublicApi.Endpoint:loop/7 | 2000 | 14767616 | 7.21 | 563.34
cowboy_clear:connection_process/4 | cowboy_http:loop/1 | 2000 | 43316544 | 21.15 | 1,652.40
Elixir.Neurow.PublicApi.SSEMonitor:init/1 | gen_server:loop/7 | 2000 | 2656000 | 1.30 | 101.32
  |   |   |   |   |  
  |   |   | Total | 29.66 | 2,317.05 |

 The subscription cost is reduced from 63ko to 30ko. In a server that handles 80k SSE subscriptions, it should save about **2.5Go of RAM**

A debounce is added to the manual garbage collection to have max 1 manual GC per SSE process every 60s. Actually the SSE passive loop is triggered much less often, the amount of manual GC should be much lower. A manual GC is always triggered during the first run of the SSE passive loop.

### CODE REVIEW

File: neurow/lib/neurow/public_api/endpoint.ex [hunk 1]
```diff
@@ -225,6 +225,8 @@ defmodule Neurow.PublicApi.Endpoint do
             jwt_exp_s
           )
 
+        :erlang.garbage_collect()
```
Comment (bpaquet) [inline neurow/lib/neurow/public_api/endpoint.ex:228]: What is the perf impact of this?

File: neurow/lib/stats.ex [hunk 2]
```diff
@@ -175,4 +175,92 @@ defmodule Neurow.Stats do
       Boolean.set([name: :stopping], true)
     end
   end
+
+  defmodule Processes do
```
[the remark this answers is not in the document]
Comment (bpaquet) [reply neurow/lib/stats.ex:179]: Yes please.

**Review by papey — COMMENTED** [review]
One thing we may want to try out is [recon](https://github.com/ferd/recon). It's a set of tool use to diagnose Erlang VM in production. 

I have tried it a bit and some the informations are interesting, eg : 

```elixir
:recon.info(self()) # self if a pid here
```

```elixir
  memory_used: [
    memory: 318128,
    message_queue_len: 0,
    heap_size: 10958,
    total_heap_size: 39648,
    garbage_collection: [
      max_heap_size: %{
        error_logger: true,
        include_shared_binaries: false,
        kill: true,
        size: 0
      },
      min_bin_vheap_size: 46422,
      min_heap_size: 233,
      fullsweep_after: 65535,
      minor_gcs: 23
    ]
  ],
  work: [reductions: 109508]
```

Here's what each part of the report represents:

- `memory_used`: This is a summary of the memory usage and related statistics for the process.

  - `memory`: The total amount of memory, in bytes, currently being used by the process.
  - `message_queue_len`: The number of messages currently in the process's message queue waiting to be processed.
  - `heap_size`: The size of the process's heap space, in words (a word is typically a few bytes depending on the architecture, for example, 4 or 8 bytes).
  - `total_heap_size`: The total size of the process's heap, including any additional overhead, in words.
  - `garbage_collection`: This contains information about the garbage collection settings and history for the process.
    - `max_heap_size`: A map that defines the behavior when the heap grows beyond a certain size.
      - `error_logger`: Whether an error log will be generated if the maximum size is reached.
      - `include_shared_binaries`: Whether the size calculation includes shared binaries.
      - `kill`: Whether the process will be killed if it exceeds the maximum heap size.
      - `size`: The maximum allowed size of the heap; a size of `0` might mean that there is no limit.
    - `min_bin_vheap_size`: The minimum virtual heap size for binaries.
    - `min_heap_size`: The minimum size of the process's heap before it starts growing.
    - `fullsweep_after`: The number of garbage collection cycles before a full sweep is triggered.
    - `minor_gcs`: The number of minor garbage collection cycles that the process has undergone.

I think it may be worth investing a bit in both `minor_gcs` and `fullsweep_after`.

`kill` was also one of you're concern @achouippe.

One [interesting read i'am digging](https://www.erlang-in-anger.com/), that mentions `recon`.

File: neurow/lib/neurow/public_api/endpoint.ex (hunk 1 above)
Comment (achouippe) [reply neurow/lib/neurow/public_api/endpoint.ex]: @bpaquet, @papey  

This is expected to be very quick because it only applied to the current process. Locally, most of the manual GC take about 6 microseconds, with a very small sets of outliers up to 20 microseconds.

Also for more context, Phoenix channels are hibernated after 15s of inactivity. Hibernating internally trigger a GC and does not seem to be a problem. So I think triggering process garbage collections is not a problem.

> One open question if the overall CPU impact on this operation distributed 10k processes.
Yes this is my main concern actually - The goal is to support about 100k per with a server with 10go RAM

I added a debounce to the manual garbage collection:
- A first GC is triggered after the SSE initialization, just after sending the history. This should be the more efficient one, because it should free all temporary resources used by cowboy and plug to initiate the HTTP connection.
- Then in the SSE passive loop, manual GC are performed max every 1 minutes when the passive loop run. (The PR does not change the frequency of the passive loop).

File: neurow/lib/stats.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (achouippe) [reply neurow/lib/stats.ex]: Done.

File: neurow/lib/neurow/observability/observability.ex [hunk 3]
```diff
@@ -0,0 +1,11 @@
+defmodule Neurow.Observability do
+  use Prometheus.Metric
+
+  def setup() do
```
Comment (papey) [inline neurow/lib/neurow/observability/observability.ex:4]: **Suggestion** 🤔 : can it be useful to create a collection of modules for observability and do something like this

```elixir
# Define a list of modules to setup
observability_modules = [
  Neurow.Observability.MessageBrokerStats,
  Neurow.Observability.HttpInterfacesStats,
  Neurow.Observability.MetricsPlugExporter,
  Neurow.Observability.SecurityStats,
  Neurow.Observability.SystemStats
]

# Iterate over the list and call the setup function on each module
Enum.each(observability_modules, fn module -> module.setup() end)
```

File: neurow/lib/neurow/observability/message_broker_stats.ex [hunk 4]
```diff
@@ -0,0 +1,82 @@
+defmodule Neurow.Observability.MessageBrokerStats do
+  use Prometheus.Metric
+
+  def setup() do
+    Gauge.declare(
```
Comment (papey) [inline neurow/lib/neurow/observability/message_broker_stats.ex:5]: **Suggestion** 🤔 : should we group the 3 "use case" in dedicated methods ?

- declare
- init
- monitor

File: neurow/lib/neurow/observability/system_stats.ex [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      message_queue_len =
+        current_stats.message_queue_len + (process_info[:message_queue_len] || 0)
+
+      Map.put(acc, {name_or_initial_func, current_func}, %ProcessesStats{
+        name_or_initial_func: name_or_initial_func,
+        process_count: process_count,
+        current_func: current_func,
+        memory: memory,
+        message_queue_len: message_queue_len
+      })
```
Comment (papey) [inline neurow/lib/neurow/observability/system_stats.ex:70]: **Suggestion** 🤔 : why not use [Map.update](https://hexdocs.pm/elixir/1.12/Map.html#update/4) here ? It takes a function that already have the previous state or default values

File: neurow/lib/neurow/observability/system_stats.ex [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    |> Enum.sort(&(&1.memory > &2.memory))
+    |> Enum.take(result_count)
+  end
+
+  defp mfa_to_string({module, function, arity}) do
+    "#{module}:#{function}/#{arity}"
+  end
+
+  defp grouping_attributes(pid) do
+    name_or_initial_func =
```
Comment (papey) [inline neurow/lib/neurow/observability/system_stats.ex:82]: **Suggestion** 🤔 : Maybe a with pattern will ease readability here

**Review by papey — APPROVED** [review]
LGTM, feedback mostly on syntax/readability

File: neurow/lib/neurow/observability/system_stats.ex (hunk 6 above)
Comment (achouippe) [reply neurow/lib/neurow/observability/system_stats.ex:82]: what do you mean ? This is already a pattern match, no ?

File: neurow/lib/neurow/observability/observability.ex (hunk 3 above)
Comment (achouippe) [reply neurow/lib/neurow/observability/observability.ex:4]: I think, for now it is easer to read without indirection.

File: neurow/lib/neurow/observability/system_stats.ex (hunk 6 above)
Comment (papey) [reply neurow/lib/neurow/observability/system_stats.ex:75]: I mean the [`with` statement](https://www.openmymind.net/Elixirs-With-Statement/),
